### PR TITLE
MEPTS 1242 | MQ CAT 13 Changes June 2023

### PR DESCRIPTION
**Pull request by TinoMuianga**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java [hunk 1]
```diff
@@ -3015,7 +3019,7 @@ public CohortDefinition getMQC13P3DEN(int indicator) {
 
     if (indicator == 2 || indicator == 9 || indicator == 10 || indicator == 11)
       compositionCohortDefinition.setCompositionString(
-          "((A AND NOT C AND NOT D AND NOT ABANDONEDTARV) OR (B1 AND NOT ABANDONED1LINE)) AND NOT (F OR E OR DD) AND age");
+          "((A AND NOT C) OR B1) AND NOT (F OR E OR DD OR ABANDONED1LINE) AND age");
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java]: Not meet requirements

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java [hunk 2]
```diff
@@ -4878,6 +4882,204 @@ public CohortDefinition getMQ13G() {
...
+    compositionCohortDefinition.addSearch(
+        "ABANDONED1LINE", EptsReportUtils.map(abandonedExclusionFirstLine, MAPPING1));
+
+    compositionCohortDefinition.addSearch(
+        "B5E",
+        EptsReportUtils.map(
+            B5E, "startDate=${startDate},endDate=${revisionEndDate},location=${location}"));
+
+    compositionCohortDefinition.setCompositionString(
+        "( B2NEW OR RESTARTED OR (B3 AND NOT B3E) ) AND NOT (ABANDONED1LINE OR B5E)");
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java]: Doesn't meet requirements

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java [hunk 3]
```diff
@@ -5820,27 +6031,30 @@ public CohortDefinition getMQ13P4(Boolean den, Integer line) {
...
-            "(((B1 AND H) AND ( (B2NEW AND NOT ABANDONEDTARV) OR (RESTARTED AND NOT (RESTARTEDTARV OR ABANDONEDTARV)) OR (B3 AND NOT B3E AND NOT (ABANDONED1LINE OR ABANDONEDTARV)) ) AND NOT B5E ) AND NOT (B4 or B5 or E or F or DD)) AND adult");
+            "(((B1 AND PrimeiraLinha AND H) AND NOT (B4 or B5 or E or F or DD)) AND adult");
       } else if (line == 12) {
         compositionCohortDefinition.setCompositionString(
-            "(((B1 AND H) AND ( (B2NEW AND NOT ABANDONEDTARV) OR (RESTARTED AND NOT (RESTARTEDTARV OR ABANDONEDTARV)) OR (B3 AND NOT B3E AND NOT (ABANDONED1LINE OR ABANDONEDTARV)) ) AND NOT B5E ) AND NOT (B4 or B5 or E or F)) AND children");
+            "(((B1 AND PrimeiraLinha AND H) AND NOT (B4 or B5 or E or F)) AND children");
       } else if (line == 18) {
         compositionCohortDefinition.setCompositionString(
-            "(((B1Den18 AND B4CV50 AND H50) AND ( (B2NEW AND NOT ABANDONEDTARV) OR (RESTARTED AND NOT (RESTARTEDTARV OR ABANDONEDTARV)) OR (B3 AND NOT B3E AND NOT (ABANDONED1LINE OR ABANDONEDTARV)) ) AND NOT B5E ) ) AND NOT (B5CV50 or E or F)");
+            "((B1Den18 AND B4CV50 AND PrimeiraLinha AND H50) AND NOT (E or F)");
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java]: Doesn't meet the requirements - B4CV50 and B1Den18

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java [hunk 4]
```diff
@@ -10256,31 +10523,51 @@ public CohortDefinition getPatientsWhoAbandonedTarvBetween3MonthsBeforePregnancy
...
+            + "                                                                 AND e.encounter_datetime <= most_recent.last_consultation  "
             + "                                                               GROUP BY p.patient_id  "
             + "                                                               UNION  "
             + "                                                               SELECT p.patient_id, max(o.obs_datetime) as last_encounter FROM patient p  "
             + "                                                                                                                                   INNER JOIN encounter e ON e.patient_id = p.patient_id  "
             + "                                                                                                                                   INNER JOIN obs o on e.encounter_id = o.encounter_id  "
-            + "                                                                                                                                   INNER JOIN (  "
-            + QualityImprovement2020Queries.getPregnancyDuringPeriod()
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java:10275]: Doesn't meet the requirements

**Review by eliazmuchave — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java (hunk 1 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java]: Solved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java (hunk 4 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java:10275]: No longer using this implementation

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java (hunk 3 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/QualityImprovement2020CohortQueries.java]: Solved

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java [hunk 5]
```diff
@@ -1174,6 +1174,115 @@ public static CohortDefinition getMQ13DenB4_P4(
...
+            + "                   FROM patient p "
+            + "						INNER JOIN encounter e ON p.patient_id = e.patient_id "
+            + "						INNER JOIN obs o ON e.encounter_id = o.encounter_id "
+            + "                   WHERE  p.voided = 0 "
+            + "                   AND e.voided = 0 "
+            + "                   AND o.voided = 0 "
+            + "                   AND o.concept_id = ${856} "
+            + "                   AND o.value_numeric >= "
+            + vlQuantity
+            + "                   AND ( e.encounter_type = ${53} AND e.encounter_datetime BETWEEN :startDate AND :endDate ) "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java]: Not obs_datetime?

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java [hunk 6]
```diff
@@ -1174,6 +1174,115 @@ public static CohortDefinition getMQ13DenB4_P4(
...
+    map.put("1982", pregnantConcept);
+    map.put("1065", yesConcept);
+
+    String query =
+        "SELECT p.patient_id "
+            + "       FROM patient p "
+            + "       INNER JOIN encounter e ON p.patient_id = e.patient_id "
+            + "       INNER JOIN obs o ON e.encounter_id = o.encounter_id "
+            + "       INNER JOIN ( "
+            + "                   SELECT p.patient_id, MIN(e.encounter_datetime) as first_carga_viral "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java]: Should select the oldest date

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java [hunk 7]
```diff
@@ -1174,6 +1174,115 @@ public static CohortDefinition getMQ13DenB4_P4(
...
+            + "				   WHERE enc.encounter_datetime = primeira.first_consultation "
+            + "                   AND enc.encounter_datetime >= :startDate "
+            + "                   AND enc.encounter_datetime <= :endDate "
+            + "                   AND enc.encounter_type = ${6} "
+            + "                   AND o.concept_id = ${1982} "
+            + "                   AND o.value_coded = ${1065} "
+            + "                   AND pe.gender = 'F' "
+            + "                   AND enc.location_id = :location "
+            + "                   GROUP BY primeira.patient_id "
+            + "                  ) AS mulher ON mulher.patient_id = p.patient_id "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java]: Should check the pregnancy concept in Ficha Resumo as well

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java [hunk 8]
```diff
@@ -1390,14 +1499,122 @@ public static CohortDefinition getMQ13DenB5_P4(
...
+            + "                                                           INNER JOIN encounter e ON p.patient_id = e.patient_id  "
+            + "                                                           INNER JOIN obs o ON e.encounter_id = o.encounter_id  "
+            + "                                                   WHERE  p.voided = 0  "
+            + "                                                    AND e.voided = 0  "
+            + "                                                    AND o.voided = 0  "
+            + "                                                    AND e.location_id = :location "
+            + "                                                    AND e.encounter_type = ${53} "
+            + "                                                    AND e.encounter_datetime <= :endDate "
+            + "                                                  GROUP BY p.patient_id "
+            + "                                          ) most_recent  ON p.patient_id = most_recent.patient_id   "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java]: Should select the record of the oldest concept, not the Ficha Resumo

**Review by eliazmuchave — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java (hunk 6 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java:1222]: Done

Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java:1222]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java (hunk 7 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/QualityImprovement2020Queries.java]: Done

**Review by eliazmuchave — APPROVED** [review]
Looks good to me

**Pull request merged**